Abstract away from language

dometto opened the pull request:
Couldn't resist this little experiment. Minor restructuring allows support for other languages than ruby (provided someone writes a parser for, e.g., python requirement files). What do you think @bartkamphorst? Most important thing is to have a clear and understandable code base, so if you think this complicates things too much we shouldn't go there.

dometto commented:
Should add: I haven't changed `app.rb` to reflect these changes. But fetching a python repo in `test.rb` works (up to the point of parsing the requirements).

In order to parse, two things are needed:
* parse the requirements file into `Gem::Requirement` objects
* provide the `PythonLang.latest_version` method (which should return a `Gem::Version` object)

So we would be using `Gem::Requirement` and `Gem::Version` as generic requirement and version classes. :)

bartkamphorst commented:
All for it! Cool idea to make this available to other languages as well. I don't really like the use of `Gem::Requirement` and `Gem::Version` that much, but maybe we can implement this functionality ourselves at some point.

dometto commented:
Finished the necessary restructuring so other languages now "work" in the app as well. Language for a project can be set in the YAML config (but if it's not set will default to `Repository::DEFAULT_LANG`, which is `'ruby'`).

bartkamphorst commented on lib/repo.rb:
```diff
@@ -60,6 +60,6 @@ def source
 
   def url_for_file(file)
-    File.join(HOST, @org, @name, @branch, file)
+    File.join(HOST, @org, @name, 'raw', @branch, file)
```
Why this literal `raw` here?

dometto replied:
That's just what a GitLab URl looks like -- github has raw.github.com, GitLab has a special route of the form 'org/name/raw/path/to/file'.

bartkamphorst commented on lib/workers/dependency.rb:
```diff
@@ -59,6 +58,7 @@ def generate_stats(locations)
       end
     end
+    results[:outdated] = :unknown if results[:outdated_total] == 0 && results[:ok] == 0
     results[:outdated] = :up_to_date unless results[:outdated]
```
Can you talk me through the logic here?

dometto replied:
If we don't manage to fetch any dependencies, all the categories will be empty. I thought it would be weird if that yielded an "all dependencies ok". So: set the status to `:unknown` if there are neither any outdated dependencies, nor any ok dependencies (but possibly unknown dependencies).

dometto commented:
Added basic support for python. Requirements files support a lot more than the basics, but [this file](https://github.com/singingwolfboy/flask-dance/blob/master/requirements.txt) (from the example project I've added to the app) already parses fine!

dometto commented:
Actually, it looks like I've pretty much got the options for `requirements.txt` covered:
* ignore lines starting with '-e' (local dependency, or dependency from versioning)
* ignore extras (e.g. `MyPackage==3.0 [PDF]`)
See https://pip.readthedocs.io/en/1.1/requirements.html

However, this only implements support for `requirements.txt`. I just noticed there is now also such a thing as a `Pipfile`: https://realpython.com/pipenv-guide/#example-usage

dometto commented on lib/langs/python.rb:
```diff
@@ -0,0 +1,66 @@
+      next if line =~ /^\s*-e/ # ignore 'editable' dependencies
+      if line =~ COMPARATORS
+        match = line.match(REGEX)
+        return nil unless match
```
Currently bailing entirely when unable to parse a single line. Does that make sense, or better to just do `next`?

bartkamphorst replied:
Either skip the line, or mark it as `:unknown`, but I prefer to keep going.

dometto commented:
Had to add Rspec to make progress developing. To do that I needed to remove the actual launching of the Sinatra app from `app.rb`, so added `bin/launch.rb` for that purpose.

Then succeeded in adding support for Pipfiles. :)

bartkamphorst commented on bin/launch.rb:
```diff
@@ -0,0 +1,66 @@
+    value.each do |name, settings|
+      cfg = settings.is_a?(Hash) ? settings : {}
+      repo = Object.const_get("#{source.capitalize}Repository").new(org, name, cfg.fetch('language', nil), cfg.fetch('locations', nil), cfg.fetch('branch', nil), cfg.fetch('api_token', nil))
+      repo.dependency_types = cfg['dependency_types'].map {|dep| dep.to_sym} if cfg['dependency_types'].is_a?(Array)
```
The `.map` call can be simplified with `.map(&:to_sym)`.

bartkamphorst replied:
Does it need to be an Array per se or can it be Array-like? Otherwise checking for Enumerator might be better.

dometto replied:
```ruby
2.6.3 :002 > [].is_a?(Enumerator)
 => false 
```
😕

bartkamphorst replied:
My bad. Try
```ruby
[].is_a?(Enumerable)
```

dometto replied:
Nice! But:

```ruby
> {}.is_a?(Enumerable)
 => true 
```
That would break, I think

bartkamphorst replied:
OK.

bartkamphorst commented on bin/launch.rb:
```diff
@@ -0,0 +1,66 @@
+include RubyDepsHelpers
+worker_fetch_all(repositories.values)
+
+sleep 5
```
This sleep still worries/bugs me. I think the way forward is to instantiate a Queue, let the DependencyWorkers dump the names and stats of processed repos into the queue (at the end of their run), and have a couple of BadgeWorkers listen for this data to come in, grab an item from the queue, and create a badge for that item.

dometto replied:
I agree and I like that idea. I'll open a separate issue for it though!

bartkamphorst commented on lib/langs/python.rb:
```diff
@@ -13,10 +17,8 @@ class PythonLang
     # Defaults for projects in this language
-    def locations(name)
+    def locations(name = nil)
       ['requirements.txt']
```
Why only requirements.txt here and not pipfile?

dometto replied:
Yeah, good question. I had assumed that the most common use case is still `requirements.txt`, as Pipfiles are pretty new. And unlike `Gemfile` and `gemspec`, a python project will use only one of them. But we might as well check for both of them by default!

dometto replied:
Added `Pipfile` now. This means, however, that we'll always be *looking* for both -- and there will, in practice, never *be* both. The risk is thus that either one of them will show up as 'missing' on the info view. Not sure what's best here.

bartkamphorst replied:
Look for one or the other, somehow? 😬

dometto replied:
The user can always set `locations: [Pipfile]` or `locations: [requirements.txt]` in the config to check for only one. But I don't see a way of automagically checking whether a project is using only the one or the other, other than fetching all locations in the worker, and just calling `.compact` on the resulting Array?

bartkamphorst replied:
OK, fair enough. Look for both by default, unless set otherwise in the config. 👍

dometto commented:
@bartkamphorst should I just merge this, or do another review?

bartkamphorst commented on lib/langs/python.rb:
```diff
@@ -56,7 +58,32 @@ def parse_requirements(content)
+
+      config.each do |type, packages|
+        packages.each do |name, info|
+          requirement = info.is_a?(String) ? info : info.fetch('version', nil)
```
An alternative here, that doesn't check for String, is this:
```ruby
requirement = info.fetch('version', nil) rescue info
```
If String coercion is important, you could change `rescue info` to `rescue info.to_s`.

bartkamphorst commented:
In the process of reviewing now.

bartkamphorst commented on lib/langs/python.rb:
```diff
@@ -67,10 +94,56 @@ def parse_pipfile(content)
       begin
-        Gem::Version.new(JSON.parse(content)['info']['version'])
-      rescue
+        v = JSON.parse(content)['info']['version']
```
version =

bartkamphorst commented on lib/langs/python.rb:
```diff
@@ -67,10 +94,56 @@ def parse_pipfile(content)
+    end
+
+    def canonical_version(v)
+      case v
```
I don't know if the case statement makes much sense here. Ternary version would work just as well:
```ruby
version.match(/^\./) ? "0#{version}" : version
```

bartkamphorst commented on lib/langs/python.rb:
```diff
@@ -67,10 +94,56 @@ def parse_pipfile(content)
+          req.sub!('.*', '.0')
+          req.sub!('!=', '')
+          v = Gem::Version.new(req) # Create the bump version using Gem::Version so pre-release handing will work
+          lower_bound = ">= #{v.bump.version}"
```
SemVer refers to `bump` as `patch` version. Perhaps we want to consistently use patch as well throughout the code base.

dometto replied:
I've changed `outdated_bump` to `outdated_patch`. (In this particular context, `bump` is the designation used by rubygems.)

bartkamphorst commented on lib/workers/helpers.rb:
```diff
@@ -4,7 +4,7 @@
 
   def log_info(message)
-    logger.info(message) if ::RubyDeps.enable_logging
+    logger.info(message) if defined?(RubyDeps) && ::RubyDeps.enable_logging
```
When would RubyDeps not be defined?

dometto replied:
I think I had this problem when setting up the tests, but I think that should work now!

bartkamphorst commented on spec/langs/python.rb:
```diff
@@ -0,0 +1,87 @@
+
+describe PythonLang do
+  context 'requirements.txt' do
+    it 'default dependency file is requirements.txt' do
```
it 'expects the default dependency file to be requirements.txt'

bartkamphorst commented on spec/langs/python_spec.rb:
```diff
@@ -0,0 +1,87 @@
+require 'langs.rb'
```
What's the difference between `spec/langs/python.rb` and `spec/langs/python_spec.rb`?

dometto replied:
Argh! It has to be called `_spec.rb`. I had first named it `python.rb`, then moved it. It happens sometimes with Sublime that you keep editing the old document, and save it. `python.rb` just needs to go. Thanks!

bartkamphorst commented on spec/langs/python_spec.rb:
```diff
@@ -0,0 +1,87 @@
+
+describe PythonLang do
+  context 'requirements.txt' do
+    it 'default dependency file is requirements.txt' do
```
Same comment as before.

dometto commented:
Getting this error on Travis, but not locally:
```
     Failure/Error: logger.error(message) if ::RubyDeps.enable_logging
```
A bit puzzled as this happens in `spec_helper.rb`:
```ruby
load File.expand_path('../../lib/app.rb', __FILE__)
RubyDeps.set(:enable_logging, false)
```
Could just add the `if defined?(::RubyDeps)`, but maybe there's another way? 😕
